### PR TITLE
implement daemonset to operator migration (storageCluster created autom…

### PR DESCRIPTION
**Pull request by haibinxie**
…atically)

Fresh operator install is not in this PR, will be implemented later.
For daemonset to operator migration, storageCluster is created automatically by operator.

### CODE REVIEW

File: charts/portworx/templates/hooks/pre-upgrade/retain-daemonset-install.yaml [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    release: {{.Release.Name | quote }}
+    chart: "{{.Chart.Name}}-{{.Chart.Version}}"
+    app.kubernetes.io/managed-by: {{.Release.Service | quote }}
+    app.kubernetes.io/instance: {{.Release.Name | quote }}
+  annotations:
+    "helm.sh/hook": pre-upgrade
+    "helm.sh/hook-weight": "10"
+    "helm.sh/hook-delete-policy": before-hook-creation
+spec:
+  {{ if semverCompare ">= 1.8" .Capabilities.KubeVersion.GitVersion }}
```
Comment (pure-jliao) [inline charts/portworx/templates/hooks/pre-upgrade/retain-daemonset-install.yaml:20]: what is this 1.8 version? k8s? how do we define this version? also how do we handle if daemonset doesn't exist? e.g. upgrading from operator 1.9

File: charts/portworx/templates/serviceaccount-hook.yaml [hunk 2]
```diff
@@ -20,12 +20,15 @@ apiVersion: {{ template "rbac.apiVersion" . }}
...
+    "helm.sh/hook": "pre-delete,post-delete,pre-upgrade"
   name: {{ template "px.hookClusterRole" . }}
 rules:
-  - apiGroups: [""]
-    resources: ["nodes"]
-    verbs: ["patch", "get", "update", "list"]
+  # for daemonset to operator migration, we need hooks for all resources deployed by daemonset, due to resources are
+  # different in different helm charts (GCP, IKS, Rancher and portworx), we use wild card here. After daemonset
+  # migration is finished for all customers we shall change this back to limited access.
+  - apiGroups: ["*"]
```
Comment (dsupure) [inline charts/portworx/templates/serviceaccount-hook.yaml:29]: We don't know when we can migrate all customers, I feel we might forget here. Can we keep the same permission and change when platform  (GCP, IKS, Rancher and portworx) needed ?

File: charts/portworx/templates/_helpers.tpl [hunk 3]
```diff
@@ -32,6 +32,18 @@ release: {{ .Release.Name | quote }}
 {{printf "%s.%s" $major $minor }}
 {{- end -}}
 
+{{- define "px.getPxOperatorImage" -}}
+{{- if (.Values.customRegistryURL) -}}
+    {{- if (eq "/" (.Values.customRegistryURL | regexFind "/")) -}}
+        {{ cat (trim .Values.customRegistryURL) "/px-operator" | replace " " ""}}
```
[the remark this answers is not in the document]
Comment (haibinxie) [reply charts/portworx/templates/_helpers.tpl:38]: That's how it is for all other images defined in px.get***Image, such as https://github.com/portworx/helm/blob/master/charts/portworx/templates/_helpers.tpl#L93

File: charts/portworx/templates/hooks/pre-upgrade/retain-daemonset-install.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (haibinxie) [reply charts/portworx/templates/hooks/pre-upgrade/retain-daemonset-install.yaml:20]: this is how it is in all other existing hooks, just to be consistent with all of them, there might be historical reasons https://github.com/portworx/helm/blob/master/charts/portworx/templates/hooks/post-install/px-create-cluster-token.yaml#L21

File: charts/portworx/templates/hooks/pre-upgrade/retain-daemonset-install.yaml [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        - name: {{ $registrySecret }}
+      {{- end }}
+      serviceAccountName: {{ template "px.hookServiceAccount" . }}
+      restartPolicy: Never
+      containers:
+        - name: retain-px-daemonset
+          {{- if eq $customRegistryURL "none" }}
+          image: "bitnami/kubectl:{{ template "px.kubectlImageTag" . }}"
+          {{- else}}
+          image: "{{ $customRegistryURL }}/bitnami/kubectl:{{ template "px.kubectlImageTag" . }}"
```
[the remark this answers is not in the document]
Comment (haibinxie) [reply charts/portworx/templates/hooks/pre-upgrade/retain-daemonset-install.yaml:38]: this is used in all other existing hooks. such as https://github.com/portworx/helm/blob/master/charts/portworx/templates/hooks/post-install/px-create-cluster-token.yaml#L37

**Review by pure-jliao — APPROVED** [review]

**Pull request closed without merging**